### PR TITLE
Fix data race in PagedArray

### PR DESCRIPTION
**Pull request by stuartcarnie**
This PR resolves a data race in the `PagedArrayPool` and `PagedArray` types. The thread sanitiser identified the data race via the `page_pool` field of the `PagedArrayPool` class. The write occurred:

https://github.com/godotengine/godot/blob/9d1cbab1c432b6f1d66ec939445bec68b6af519e/core/templates/paged_array.h#L62

and the concurrent read here:

https://github.com/godotengine/godot/blob/9d1cbab1c432b6f1d66ec939445bec68b6af519e/core/templates/paged_array.h#L78

The actual data race of the `page_pool` pointer being access concurrently is possibly safe, as it is a word read / write, however, an additional problem is the underlying memory in the `memrealloc` from the old allocation is being immediately zero'd in debug builds on macOS, so the call the `get_page` is sometimes returning an `nullptr`. In a release build, the old allocation may be immediately reused before the call to `get_page`, returning an invalid pointer, resulting in crashes, memory corruption, etc.

### CODE REVIEW

File: core/templates/paged_array.h [hunk 1]
```diff
@@ -53,7 +53,12 @@ class PagedArrayPool {
 	SpinLock spin_lock;
 
 public:
-	uint32_t alloc_page() {
+	struct PageInfo {
+		T *page;
```
Comment (RandomShaper) [inline core/templates/paged_array.h:57]: I'd suggest adding initializers. That's the trend in Godot lately. I think not adding them should be the exception (for spots where perf is critical and the compiler can't be trusted to optimize out redundant assignments).

**Review by RandomShaper — CHANGES_REQUESTED** [review]
Looks good besides the little nitpick.

File: core/templates/paged_array.h (hunk 1 above)
Comment (stuartcarnie) [reply core/templates/paged_array.h:57]: Sounds good 👍🏻

**Review by RandomShaper — APPROVED** [review]

**Comment by akien-mga** [comment]
Looks good! Could you squash the commits? See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) for instructions.

**Comment by stuartcarnie** [comment]
@RandomShaper squashed

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada: -- hopefully the start of many, given what I know you've been cooking ;)

**Comment by akien-mga** [comment]
Cherry-picked for 4.2.2.

**Pull request merged**